### PR TITLE
Autofill index prefix from title in index set creation

### PR DESCRIPTION
**Pull request by boosty**
## Description

When creating a new index set, the index prefix field now automatically populates based on the title field to improve user experience. The prefix is generated by converting the title to lowercase and replacing special characters with hyphens. Autofill stops when users manually edit the prefix field. This feature only applies during creation since the prefix is immutable after an index set is created.

## Motivation and Context

When entering a title for an index set, save users some typing by generating an index prefix for them. 

Closes https://github.com/Graylog2/graylog2-server/issues/24568

## How Has This Been Tested?

- Added test cases in `IndexSetConfigurationForm.test.tsx` and `transformTitleToPrefix.test.ts`
- Manual testing in Safari and Chrome

## Screenshots (if appropriate):

Screenshot 2025-12-23 at 14 53 18

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx [hunk 1]
```diff
@@ -187,6 +188,32 @@ const IndexSetConfigurationForm = ({
...
   };
 
+  const transformTitleToPrefix = (title: string): string => {
+    if (!title) return '';
+
+    return title
+      .trim()
+      .toLowerCase()
+      .replace(/[^a-z0-9_+-]+/g, '-') // Replace invalid chars with hyphen
+      .replace(/^[_+-]+/, '') // Remove leading invalid chars
```
[the remark this answers is not in the document]
Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: I think it does not make sense to document the whole regexp as plain text. The tests and help text for the input field should suffice.

[the remark this answers is not in the document]
Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: Same here. I think it does not make sense to document the whole regexp as plain text. The tests and help text for the input field should suffice.

File: graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx [hunk 2]
```diff
@@ -187,6 +188,32 @@ const IndexSetConfigurationForm = ({
     setFieldTypeRefreshIntervalUnit(unit);
   };
 
+  const transformTitleToPrefix = (title: string): string => {
+    if (!title) return '';
+
+    return title
+      .trim()
+      .toLowerCase()
+      .replace(/[^a-z0-9_+-]+/g, '-') // Replace invalid chars with hyphen
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: Maybe use lodash's [deburr](https://lodash.com/docs/4.17.15#deburr) before this, so we do not replace extended letters with `-`?

Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: That sounds good! I will have a look.

Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: @dennisoelkers Added in 56fc91d72d6556bf163b3bf198ece0f5e61c7254

Comment (dennisoelkers) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: That is great! Can we move the function out of the component function and export it? Because:

1.) It is a pure function and should not be declared in the component's render path
2.) We can write/generate a unit test for it covering all of the edge cases

**Review by dennisoelkers — COMMENTED** [review]
Thanks for this! It looks and works great. I just added one minor suggestion inline. If you have the time to address it, that would be great. If not, we could also merge it as-is.

File: graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx [hunk 3]
```diff
@@ -78,6 +79,18 @@ const SubmitWrapper = styled.div`
...
 
+export const transformTitleToPrefix = (title: string): string => {
+  if (!title) return '';
+
+  return deburr(title) // Replace extended letters with basic Latin letters
+    .trim()
+    .toLowerCase()
+    .replace(/[^a-z0-9_+-]+/g, '-') // Replace invalid chars with hyphen
+    .replace(/^[_+-]+/, '') // Remove leading invalid chars
+    .replace(/-+$/, '') // Remove trailing hyphens
```
[the remark this answers is not in the document]
Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx:90]: Thought about that too, but I think it is ok to keep it as is. The reason for just removing the hyphen at the end is that the hyphen is our replacement for invalid characters, so our function might have put it there—and that one looks superfluous.

File: graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.test.tsx [hunk 4]
```diff
@@ -288,4 +290,42 @@ describe('IndexSetConfigurationForm', () => {
...
+
+    await waitFor(() => {
+      expect(prefixInput).toHaveValue('my-test-index');
+    });
+
+    // Test autofill stops after manual edit
+    userEvent.clear(prefixInput);
+    userEvent.type(prefixInput, 'custom-prefix');
+    userEvent.clear(titleInput);
+    userEvent.type(titleInput, 'Another Title');
```
[the remark this answers is not in the document]
Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.test.tsx]: Added in e867d9a

File: graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx (hunk 2 above)
Comment (boosty) [reply graylog2-web-interface/src/components/indices/IndexSetConfigurationForm/IndexSetConfigurationForm.tsx]: Good idea! I changed it as suggested and added the corresponding unit test.

**Comment by boosty** [comment]
@dennisoelkers Thanks a lot for the thorough review so far! I think I have addressed all points now. Please have another look  when you have time.
